File: server/gemini_service.ts
import { GoogleGenAI, Type } from "@google/genai";

let ai: GoogleGenAI | null = null;

function getGeminiClient(): GoogleGenAI | null {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey || apiKey === "MY_GEMINI_API_KEY" || apiKey === "YOUR_GEMINI_API_KEY" || apiKey.trim() === "") {
    console.warn("GEMINI_API_KEY is not defined or is set to a placeholder. Using local fallback analysis.");
    return null;
  }
  if (!ai) {
    ai = new GoogleGenAI({
      apiKey,
      httpOptions: {
        headers: {
          'User-Agent': 'aistudio-build',
        }
      }
    });
  }
  return ai;
}

export interface GeminiAnalysis {
  video_summary: string;
  topic_tags: string[];
  primary_topic: string;
  content_category: string;
  knowledge_domain: string;
  content_type: string;
  estimated_user_intent: string;
  emotion_tone: string;
  confidence_score: number;
}

export async function analyzeVideoMetadata(metadata: any): Promise<GeminiAnalysis> {
  const client = getGeminiClient();
  if (!client) {
    return getFallbackAnalysis();
  }

  const desc = metadata.description ? metadata.description.slice(0, 1000) : "";
  const prompt = `
Analyze the following YouTube video metadata and estimate its semantic and behavioral characteristics.

Video ID: ${metadata.video_id || ""}
Title: ${metadata.official_title || ""}
Description: ${desc}
Channel Name: ${metadata.channel_name || ""}
Category ID: ${metadata.category_id || ""}
Tags: ${metadata.tags && metadata.tags.length > 0 ? metadata.tags.join(", ") : "None"}

You must return a valid, minified JSON object with the following fields and constraints:
1. "video_summary": (string, max 3-sentence summary emphasizing semantic content, educational/entertainment value, or cognitive state/intent of someone watching)
2. "topic_tags": (array of 3-5 specific semantic topic tags or keywords)
3. "primary_topic": (string, the single most prominent topic of the video)
4. "content_category": (string, must be exactly one of: Educational, Entertainment, News/Documentary, Tech/Gaming, Self-Improvement/Lifestyle, Music/Art, Other)
5. "knowledge_domain": (string, broad academic or professional domain, e.g. Computer Science, Philosophy, Pop Culture, Fitness, Finance, etc.)
6. "content_type": (string, must be exactly one of: Long-form, Short-form, Lecture, Review, Gameplay, Music Video, Vlog, Essay, Other)
7. "estimated_user_intent": (string, must be exactly one of: Learning, Entertainment, Information_Seeking, Skill_Acquisition, Escapism, Relaxation, Curiosity, Other)
8. "emotion_tone": (string, comma-separated list of 1-3 dominant emotional or mental tones, e.g. Calming, Exciting, Analytical, Humorous, Inspiring, Tense, Academic, Casual, etc.)
9. "confidence_score": (float between 0.0 and 1.0 representing confidence in this classification)

Your response must contain ONLY the valid JSON object. No markdown formatting, no code blocks (do not wrap in \`\`\`json), no additional text.
`;

  try {
    const response = await client.models.generateContent({
      model: "gemini-3.5-flash",
      contents: prompt,
      config: {
        responseMimeType: "application/json",
        responseSchema: {
          type: Type.OBJECT,
          properties: {
            video_summary: { type: Type.STRING },
            topic_tags: { type: Type.ARRAY, items: { type: Type.STRING } },
            primary_topic: { type: Type.STRING },
            content_category: { type: Type.STRING },
            knowledge_domain: { type: Type.STRING },
            content_type: { type: Type.STRING },
            estimated_user_intent: { type: Type.STRING },
            emotion_tone: { type: Type.STRING },
            confidence_score: { type: Type.NUMBER },
          },
          required: [
            "video_summary",
            "topic_tags",
            "primary_topic",
            "content_category",
            "knowledge_domain",
            "content_type",
            "estimated_user_intent",
            "emotion_tone",
            "confidence_score"
          ]
        }
      }
    });

    let text = response.text || "";
    text = text.trim();
    if (text.startsWith("```")) {
      const lines = text.split("\n");
      if (lines[0].startsWith("```")) {
        lines.shift();
      }
      if (lines[lines.length - 1].startsWith("```")) {
        lines.pop();
      }
      text = lines.join("\n").trim();
    }

    return JSON.parse(text) as GeminiAnalysis;
  } catch (err: any) {
    console.warn(
      `Gemini API key is invalid or not active yet. Falling back to local/simulated analysis. ` +
      `Ensure you have provided a valid GEMINI_API_KEY in the Settings > Secrets panel of your AI Studio workspace. ` +
      `Error details: ${err.message || err}`
    );
    return getFallbackAnalysis();
  }
}

export function getFallbackAnalysis(): GeminiAnalysis {
  return {
    video_summary: "Analysis unavailable. Fallback generated due to processing limitation.",
    topic_tags: ["general", "uncategorized"],
    primary_topic: "General Content",
    content_category: "Other",
    knowledge_domain: "General",
    content_type: "Other",
    estimated_user_intent: "Curiosity",
    emotion_tone: "Neutral",
    confidence_score: 0.5
  };
}

File: gemini_service.py
import os
import json
import asyncio
from google import genai
from google.genai import types

def get_gemini_client():
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        print("GEMINI_API_KEY is not defined. Using fallback analysis.")
        return None
    try:
        # User-Agent config is set via Client
        return genai.Client(api_key=api_key, http_options={"headers": {"User-Agent": "aistudio-build"}})
    except Exception as e:
        print(f"Failed to initialize Gemini Client: {e}")
        return None

async def analyze_video_metadata(metadata: dict) -> dict:
    client = get_gemini_client()
    if not client:
        return get_fallback_analysis(metadata)

    video_id = metadata.get("video_id", "")
    title = metadata.get("official_title", "")
    desc = metadata.get("description", "")
    if desc:
        desc = desc[:1000]
    channel = metadata.get("channel_name", "")
    cat_id = metadata.get("category_id", "")
    tags = metadata.get("tags", [])

    prompt = f"""
Analyze the following YouTube video metadata and estimate its semantic and behavioral characteristics.

Video ID: {video_id}
Title: {title}
Description: {desc}
Channel Name: {channel}
Category ID: {cat_id}
Tags: {", ".join(tags) if tags else "None"}

You must return a valid, minified JSON object with the following fields and constraints:
1. "video_summary": (string, max 3-sentence summary emphasizing semantic content, educational/entertainment value, or cognitive state/intent of someone watching)
2. "topic_tags": (list of 3-5 specific semantic topic tags or keywords)
3. "primary_topic": (string, the single most prominent topic of the video)
4. "content_category": (string, must be exactly one of: Educational, Entertainment, News/Documentary, Tech/Gaming, Self-Improvement/Lifestyle, Music/Art, Other)
5. "knowledge_domain": (string, broad academic or professional domain, e.g. Computer Science, Philosophy, Pop Culture, Fitness, Finance, etc.)
6. "content_type": (string, must be exactly one of: Long-form, Short-form, Lecture, Review, Gameplay, Music Video, Vlog, Essay, Other)
7. "estimated_user_intent": (string, must be exactly one of: Learning, Entertainment, Information_Seeking, Skill_Acquisition, Escapism, Relaxation, Curiosity, Other)
8. "emotion_tone": (string, comma-separated list of 1-3 dominant emotional or mental tones, e.g. Calming, Exciting, Analytical, Humorous, Inspiring, Tense, Academic, Casual, etc.)
9. "confidence_score": (float between 0.0 and 1.0 representing confidence in this classification)

Your response must contain ONLY the valid JSON object. No markdown formatting, no code blocks (do not wrap in ```json), no additional text.
"""
    # Valid candidate models in order of preference according to AI Studio guidelines
    candidate_models = ['gemini-3.6-flash', 'gemini-flash-latest', 'gemini-3.1-flash-lite']

    for model_name in candidate_models:
        for attempt in range(2):
            try:
                response = client.models.generate_content(
                    model=model_name,
                    contents=prompt,
                    config=types.GenerateContentConfig(
                        response_mime_type="application/json",
                    ),
                )
                text = response.text.strip()
                
                # Clean up any potential markdown wrapper just in case
                if text.startswith("```"):
                    lines = text.split("\n")
                    if lines[0].startswith("```"):
                        lines = lines[1:]
                    if lines[-1].startswith("```"):
                        lines = lines[:-1]
                    text = "\n".join(lines).strip()
                    
                parsed = json.loads(text)
                if isinstance(parsed, dict) and "video_summary" in parsed:
                    return parsed
            except Exception as e:
                print(f"Gemini API call ({model_name}, attempt {attempt+1}) failed: {e}")
                if attempt < 1:
                    await asyncio.sleep(1)

    print("All Gemini models failed or busy. Using smart fallback analysis.")
    return get_fallback_analysis(metadata)

def get_fallback_analysis(metadata: dict = None) -> dict:
    if not metadata:
        metadata = {}
    title = metadata.get("official_title", "").strip() or "Untitled Video"
    channel = metadata.get("channel_name", "").strip()
    tags = metadata.get("tags", [])
    
    # Infer basic category heuristic based on title keywords
    title_lower = title.lower()
    category = "Other"
    intent = "Curiosity"
    domain = "General"
    
    if any(k in title_lower for k in ["tutorial", "how to", "guide", "explained", "learn", "course", "lecture"]):
        category = "Educational"
        intent = "Learning"
        domain = "Education & Tech"
    elif any(k in title_lower for k in ["review", "test", "vs", "code", "python", "javascript", "react", "tech"]):
        category = "Tech/Gaming"
        intent = "Skill_Acquisition"
        domain = "Technology"
    elif any(k in title_lower for k in ["song", "music", "audio", "remix", "official video", "band", "album"]):
        category = "Music/Art"
        intent = "Relaxation"
        domain = "Music"
    elif any(k in title_lower for k in ["funny", "vlog", "gameplay", "stream", "comedy", "movie"]):
        category = "Entertainment"
        intent = "Escapism"
        domain = "Pop Culture"

    topic_tags = [t for t in tags[:4]] if tags else [category.lower(), "video"]
    if not topic_tags:
        topic_tags = ["general"]

    summary_part = f"Video by {channel} titled '{title}'." if channel else f"Video titled '{title}'."

    return {
        "video_summary": f"{summary_part} Provides content focused on {domain.lower()} concepts.",
        "topic_tags": topic_tags,
        "primary_topic": title[:50] if title else "General Content",
        "content_category": category,
        "knowledge_domain": domain,
        "content_type": "Long-form",
        "estimated_user_intent": intent,
        "emotion_tone": "Informative, Casual",
        "confidence_score": 0.7
    }

